### DISCUSSION TITLE
Interoperability: How to convert Spectre data.table to object compatible with CATALYST package?

### DISCUSSION THREAD

**Discussion by denvercal1234GitHub** [ImmuneDynamics/Spectre#92] [Q&A] [answered]
Hi there, 

I lov Spectre package especially with the well-crafted tutorials and explanation. But, I also hope to use some features provided by the CATALYST package. 

CATALYST package organizes their data as a SingleCellExperiment class (https://bioconductor.org/packages/release/bioc/vignettes/CATALYST/inst/doc/differential.html).

Do you have a recommended way to do this?

Thank you for your help!

**Comment by tomashhurst** [comment]
@denvercal1234GitHub it's pretty straightforward to construct a SingleCellExperiment from a `data.table`, and something I have been meaning to create a quick tutorial for. @ghar1821 would you have time to write up a quick example?

**Reply by ghar1821** [reply]
Yep I'll have time to write up a quick tutorial. I'll post a link here when I've written it up!

**Reply by denvercal1234GitHub** [reply]
Thank you so much, @ghar1821 !

**Reply by denvercal1234GitHub** [reply]
hI @ghar1821 - Would you mind letting me know whether you had gotten a chance to write the tutorial for converting Spectre data.table to SCE object? Thank you so much!

**Reply by ghar1821** [reply]
@denvercal1234GitHub 
I've written up a full tutorial here: https://immunedynamics.io/spectre/tutorials/datatable_interoperability/sce_support.html

Let me know if you need help.

**Comment by ghar1821** [comment, marked as answer]
@denvercal1234GitHub 

Sorry for the huge delay! I haven't written a full tutorial, but I'll attach a brief one here which should help:

With sce, cell-specific (e.g., what sample or group each cell belongs to) or marker-specific (e.g., what are the channels) must be separate from the actual data. 
See the following where I mock up my data and store it as `dt`, and the cell- and marker-specific information as `cell_info` and `marker_info`.
```
dt <- data.table(
    CD4=sample(1:100, 20, replace = TRUE),
    CD8=sample(1:100, 20, replace = TRUE),
    CD3=sample(1:100, 20, replace = TRUE))

marker_info <- data.table(
    name=c("CD4", "CD8", "CD3"),
    type=rep("Cell type", 3)
)

cell_info <- data.table(
    cell_id=paste0("Cell_", c(1:20)),
    sample="01",
    group="healthy"
)
```

Next, to load them up into sce, we start with just the data, like the following:

```
library(SingleCellExperiment)
sce <- SingleCellExperiment(assays=list(counts=t(dt)))
```

Note a couple of things:
1. We store the data in an assay call count in sce.
2. I transformed the data so the cells are columns, and markers are row. This is the standard for sce and scRNAseq data.

Then we load up the marker- and cell- specific information as rowData and colData respectively, while also giving the row and column in the sce object a unique id. For markers (rows), it can just be the name, for cells (columns), you will probably have to make it up using something like `paste0("CellId"_c(1:nrow(dt)))` if you don't have anything that uniquely identifies each cell.

```
# Load up the marker information.
# Unique name for each row seem to already be loaded from previous step. 
rowData(sce) <- DataFrame(marker_info, row.names = marker_info$name)

# Give each cell a unique ID, already embedded in my original data table
colnames(sce) <- cell_info$cell_id
# Then load up the rest of the information
colData(sce) <- DataFrame(cell_info, row.names = cell_info$cell_id)
```

Check your object:
```
sce
```

and it should look like this:
```
class: SingleCellExperiment 
dim: 3 20 
metadata(0):
assays(1): counts
rownames(3): CD4 CD8 CD3
rowData names(2): name type
colnames(20): Cell_1 Cell_2 ... Cell_19 Cell_20
colData names(3): cell_id sample group
reducedDimNames(0):
mainExpName: NULL
altExpNames(0):
```

Hope this help! Let me know if you get stuck or need more info. I'll write up a proper tutorial in the coming days.
